### ISSUE TITLE
support JQ for API response transformation

### ISSUE THREAD

**Issue by newgene** [biothings/biothings_explorer#489]
[JMESPath ](https://jmespath.org/)is a query language for JSON, with [libraries available in many languages](https://jmespath.org/libraries.html) (e.g. javascript and Python).

We can evaluate and consider to use it in [api-respone-transform.js](https://github.com/biothings/api-respone-transform.js) sub-module. Potentially remove or simplify [the existing hard-coded transformers](https://github.com/biothings/api-respone-transform.js/tree/main/src/transformers). 

The overall benefits I believe this new feature can bring us:
* Allow to standardize the API response transformation in a well-defined language, and we can then put it into the SmartAPI metadata, instead of hard-coded API-specific code within the module.
* Simplify the complexity of the [api-respone-transform.js](https://github.com/biothings/api-respone-transform.js) sub-module, since the many transformation logic will be handled by JMESPath module.
* Support more complex transformation logic if needed.

**Comment by newgene** [comment]
`jq` can be another option, but `jq` module is more commandline-focused, while `JMESPath` is more suitable to be used in our modules.

**Comment by colleenXu** [comment]
how does this relate to the templating work (https://github.com/biothings/call-apis.js/pull/30, https://github.com/biothings/call-apis.js/issues/26, https://github.com/biothings/call-apis.js/pull/31)?

Also is there any concern about SmartAPI x-bte annotation being complex / very code-like @andrewsu ?

**Comment by rjawesome** [comment]
@ericz1803 if you have other stuff to work on right now, I can take this issue.

**Comment by ericz1803** [comment]
Yeah you can go ahead and take it

**Comment by rjawesome** [comment]
I started work by trying it on one transformer on the use-jmes-path branch of api-response-transform. For now I'm trying JQ because it seems jmes path is missing some functions like string splitting.

**Comment by rjawesome** [comment]
I created some jq transformers for EBI, CTD, and OpenTarget. Is there any test queries I can use for any of these?

**Comment by andrewsu** [comment]
Next steps in this issue: 
* @newgene to reach out to @rjawesome to discuss jq vs JMESPath
* finish migrating the rest of the transformers

Future issues:
* move jq code to SmartAPI metadata
* update documentation describing necessary formats
* update BTE to consume jq annotation

**Comment by colleenXu** [comment]
Replying to Rohan's [comment](https://github.com/biothings/BioThings_Explorer_TRAPI/issues/489#issuecomment-1239747435):
* I think I removed CTD from BTE's list of APIs a while back due to an outdated SmartAPI yaml. 
* I [removed](https://github.com/biothings/BioThings_Explorer_TRAPI/issues/432#issuecomment-1109324120) OpenTargets from BTE's list of APIs more recently because that API had been removed and replaced with a diff query-interface. 

I think the EBI transformer is for EBI Proteins API? So the test-query I have for that API is this (from the [comments](https://github.com/NCATS-Tangerine/translator-api-registry/blob/master/ebi_proteins/openapi.yml#L3244) in the x-bte section):

```
{
    "message": {
        "query_graph": {
            "edges": {
                "e01": {
                    "subject": "n0",
                    "object": "n1"
                }
            },
            "nodes": {
                "n0": {
                    "ids": ["UniProtKB:Q93099", "UniProtKB:Q9Y6P5"],
                    "categories": ["biolink:Gene"]
                },
                "n1": {
                    "categories": ["biolink:MolecularActivity"]
                }
            }
        }
    }
}
```

**Comment by colleenXu** [comment]
JQ draft PR https://github.com/biothings/api-respone-transform.js/pull/38

**Comment by colleenXu** [comment]
Notes from 2023-03-22 group meeting:

This issue is related to post-query processing, which is especially important for BTE correctly handling APIs "in the wild" / external APIs (specifically this means non-TRAPI / non-BioThings). 

What is in the JQ PR? 
1) replaced some custom js with jq processing. Some jq processing can stay in this repo as reusable functions for SmartAPI yamls 
2) some jq processing can be taken out and moved into the individual SmartAPI yamls 
3) custom js code is still the fallback for post-query processing

Agreement: let’s move forward with getting the first JQ PR https://github.com/biothings/api-respone-transform.js/pull/38 incorporated / deployed. 

---

Plan (sequential):

- JC to check PR: actually uses JQ processing / update by merging latest main
- CX to test existing JQ processing (code)
    - biolink/monarch
    - Ebi proteins
    - CTD
    - Semmeddb / biothings?? Ask Rohan for clarification
- JC deploy to dev
- Coordinate:
    - CX does (2) above. Maybe adds more jq stuff (CTD/RaMP?). Related to https://github.com/biothings/BioThings_Explorer_TRAPI/issues/521
    - CX / JC remove jq from code
    - CX tests, JC tests, work towards deploying it to all instances

Didn’t have time to review use-cases and ask whether JQ processing can handle it. That’s fine, since we’ve decided to incorporate it anyways
* [CTD use-cases](https://github.com/biothings/BioThings_Explorer_TRAPI/issues/558#issuecomment-1467465374): [1](https://github.com/biothings/BioThings_Explorer_TRAPI/issues/583), [2](https://github.com/biothings/BioThings_Explorer_TRAPI/issues/584), [3](https://github.com/biothings/BioThings_Explorer_TRAPI/issues/585), [4](https://github.com/biothings/BioThings_Explorer_TRAPI/issues/586)
* [RaMP use-cases](https://github.com/biothings/pending.api/issues/69#issuecomment-1473375927) (3, one is non-overlapping with CTD)

**Comment by colleenXu** [comment]
### Notes from today's group meeting

Design decision / plan: Shouldn't eliminate any existing feature, instead adds choice 

#### API-post-processing situations

* JS in BTE: CW,JC seem fairly sure JS is more performant than JQ, so we want to keep using JS. Let's keep this for internal APIs (BioThings), very common transforms, or complex transforms 
* JQ in BTE: lots of this in the current JQ PRs, as proof-of-concept. Not sure how much we want to actually use this - maybe only for "commonly used" JQ stuff? 
* JQ in SmartAPI yaml: Use JQ in SmartAPI yaml when possible for external APIs. It's particuarly good for simple transforms (parsing single fields?) and api-specific transforms 
    * allows those APIs' teams to help build/maintain post-processing "code", rather than have our team do this
    * ideally more transparent in this way 

If we notice stuff used in a lot of post-processing, we can keep it in the BTE module. We can also improve the Base Transformer class (would be a separate issue) 

#### → next steps are JC and CX working to review the current PRs to see what to "not use" (use JS instead), what to keep as "JQ in BTE", and what to move to SmartAPI yaml (see post-processing situations above)

**Comment by colleenXu** [comment]
### Notes from Jackson and me's meeting today

Jackson will fix the typo "jq_transfomer" -> file name and all refs (and check smartapi-kg.js for this typo)

We discussed what APIs are covered by the current JQ work, and "where is extra JQ functionality (not covered by JS in main branch)"? And what to do next?

#### BioThings:
* [handling more complex "q" and pairing input IDs with hits](https://github.com/biothings/biothings_explorer/issues/592)
* (note that [list_filter_all and list_filter_any JQ work](https://github.com/biothings/api-respone-transform.js/pull/38#issuecomment-1381186681) was related to [BioThings API stuff and EBI Protein stuff](https://github.com/biothings/api-respone-transform.js/pull/38#issuecomment-1374342469))

➡️  ~agreed to move any extra functionality to JS. Want to use JS instead for performance reasons and because many of our APIs are BioThings (internal).~

9/18: @tokebe suggested keeping this JQ. So next is checking that the JQ stuff, including this "extra functionality", works. Dunno about moving to SmartAPI yaml - will discuss later

#### EBI Proteins

added extra functionality in JQ string, described [here](https://github.com/biothings/api-respone-transform.js/pull/38#issuecomment-1371259413) and [here](https://github.com/biothings/api-respone-transform.js/pull/38#issuecomment-1374342469)

➡️  check that this JQ stuff works, and move to SmartAPI yaml. 

#### CTD

* [fields with pipe-delimited values](https://github.com/biothings/biothings_explorer/issues/583#issuecomment-1469294510) - overlap with this [JS quick-fix](https://github.com/biothings/api-respone-transform.js/pull/53#issuecomment-1674285164)?
* [batch-query support](https://github.com/biothings/biothings_explorer/issues/584#issuecomment-1471197578), comment [here](https://github.com/biothings/api-respone-transform.js/pull/38#issuecomment-1480444912)
    * I notice that RaMP has a similar issue, called "ISSUE 2" in these [two](https://github.com/biothings/pending.api/issues/69#issuecomment-1473375927) [posts](https://github.com/biothings/pending.api/issues/69#issuecomment-1474740449). However, working on RaMP-related post-processing issues may be out-of-scope for now...

➡️   check that this JQ stuff works, and move to SmartAPI yaml. 

#### Biolink/monarch

Didn't notice any extra functionality ([note that JQ is being used](https://github.com/biothings/api-respone-transform.js/pull/38#issuecomment-1266043571))

➡️  check that this JQ stuff works, and move to SmartAPI yaml.

#### Overall

* If we want to keep JQ-processing, does this work okay with the publication refactor?
* There's other external APIs, not sure if they're covered by current JQ work or if we will want JQ stuff for them in the future:
    * litvar
    * quickgo
    * ols

**Comment by colleenXu** [comment]
(my report part 1, pasted from Slack DMs with Jackson)

I highlighted one bug with CTD for @tokebe to look into...

> Done reviewing "potentially-extra-functionality to JQ" stuff. Didn't find anything more - they were just tagged "we could try seeing if this can be addressed with JQ, once it is deployed :stuck_out_tongue: "
>
> Also tested current JQ PR (just the api-response-transform). I couldn't figure out how to successfully test w/ breakpoints, so I'm not 100% sure the JQ code is being used. For EBI Proteins, Biolink/Monarch seemed to work as-intended, YAY.
>
> But for CTD:
> * I think there's a bug when using [this operation](https://github.com/NCATS-Tangerine/translator-api-registry/blob/bdd32698844f5dfd29758a7700b6952a8e206e39/CTD/smartapi.yaml#L516). Running the testExample: Pathway KEGG.PATHWAY:hsa05323 -> Gene...89 records are retrieved but they're all dropped during edge-management...so no results are returned. I think this operation previously worked...
> * worked as-intended for all other operations
>  * worked as-intended for extra functionality: [pipe-delimited](https://github.com/biothings/biothings_explorer/issues/583#issuecomment-1469294510), and [batch-querying](https://github.com/biothings/biothings_explorer/issues/584)

**Comment by colleenXu** [comment]
(my report part 2, updated version of what's on Slack)

@tokebe 

It also seems like the BioThings APIs [extra functionality](https://github.com/biothings/biothings_explorer/issues/592) in JQ isn't working...

To test this, you'll need x-bte annotation that is supposed to use this. You can take the [SuppKG yaml](https://github.com/NCATS-Tangerine/translator-api-registry/blob/bdd32698844f5dfd29758a7700b6952a8e206e39/suppkg/suppkg.yaml#L581), replace the operations (line 581-630) with the stuff below, and set up an override on your local BTE.

<details><summary>SuppKG operations that should use this functionality by putting the input ID in the middle of the query-string</summary>
<p>

```
    supplement-treats-disease:
    ## 595,222 records 
      - supportBatch: true
        useTemplating: true ## flag to say templating is being used below
        inputs:
          - id: UMLS
            semantic: SmallMolecule
        requestBodyType: object
        requestBody:
          body: >-
            {"q": {{ queryInputs | wrap('subject.umls:', ' AND predicate:TREATS
            AND object.semtypes:(acab OR anab OR cgab OR comd OR dsyn OR mobd OR neop)') | dump }}, "scopes": []}
        outputs:
          - id: UMLS
            semantic: Disease
        parameters:
          fields: object.umls,relation.pmid,relation.sentence,subject.name,object.name
          size: 1000
        predicate: treats
        source: "infores:suppkg" # no infores for suppkg yet
        response_mapping:
          "$ref": "#/components/x-bte-response-mapping/object"
        # testExamples:
        #   - qInput: "UMLS:C0016277"      ## Fluconazole
        #     oneOutput: "UMLS:C0009324"   ## ulcerative colitis
    disease-treated_by-supplement:
      - supportBatch: true
        useTemplating: true
        inputs:
          - id: UMLS
            semantic: Disease
        requestBodyType: object
        requestBody:
          body: >-
            {"q": {{ queryInputs | wrap('object.umls:', ' AND predicate:TREATS AND (NOT subject.semtypes:(bact OR gngm OR plnt))
            AND subject.semtypes:(aapp OR antb OR bacs OR dsp OR food OR inch OR orch OR phsu OR vita)') | dump }}, "scopes": []}
        outputs:
          - id: UMLS
            semantic: SmallMolecule
        parameters:
          fields: subject.umls,relation,subject.name,object.name
          size: 1000
        predicate: treated_by
        source: "infores:suppkg" # no infores for suppkg yet
        response_mapping:
          "$ref": "#/components/x-bte-response-mapping/subject"
        # testExamples:
        #   - qInput: "UMLS:C0009324"      ## ulcerative colitis
        #     oneOutput: "UMLS:DC0023791"   ## (r)-dithiolane-3-pentanoic acid
```

</p>
</details> 

<details><summary>What I see in my local logs</summary>
<p>

It looks like BTE isn't parsing the api-response into records successfully...

```
  bte:call-apis:query using template builder +45s
  bte:call-apis:query {
  bte:call-apis:query   url: 'https://biothings.ncats.io/suppkg/query',
  bte:call-apis:query   params: {
  bte:call-apis:query     with_total: true,
  bte:call-apis:query     fields: 'object.umls,relation.pmid,relation.sentence,subject.name,object.name',
  bte:call-apis:query     size: 1000
  bte:call-apis:query   },
  bte:call-apis:query   data: {
  bte:call-apis:query     q: [
  bte:call-apis:query       'subject.umls:C0016277 AND predicate:TREATS AND object.semtypes:(acab OR anab OR cgab OR comd OR dsyn OR mobd OR neop)',
  bte:call-apis:query       'subject.umls:C4082694 AND predicate:TREATS AND object.semtypes:(acab OR anab OR cgab OR comd OR dsyn OR mobd OR neop)',
  bte:call-apis:query       'subject.umls:C5164749 AND predicate:TREATS AND object.semtypes:(acab OR anab OR cgab OR comd OR dsyn OR mobd OR neop)',
  bte:call-apis:query       'subject.umls:C5164750 AND predicate:TREATS AND object.semtypes:(acab OR anab OR cgab OR comd OR dsyn OR mobd OR neop)',
  bte:call-apis:query       'subject.umls:C5164751 AND predicate:TREATS AND object.semtypes:(acab OR anab OR cgab OR comd OR dsyn OR mobd OR neop)',
  bte:call-apis:query       'subject.umls:C5164754 AND predicate:TREATS AND object.semtypes:(acab OR anab OR cgab OR comd OR dsyn OR mobd OR neop)',
  bte:call-apis:query       'subject.umls:C5164752 AND predicate:TREATS AND object.semtypes:(acab OR anab OR cgab OR comd OR dsyn OR mobd OR neop)',
  bte:call-apis:query       'subject.umls:C5164753 AND predicate:TREATS AND object.semtypes:(acab OR anab OR cgab OR comd OR dsyn OR mobd OR neop)',
  bte:call-apis:query       'subject.umls:C4266171 AND predicate:TREATS AND object.semtypes:(acab OR anab OR cgab OR comd OR dsyn OR mobd OR neop)',
  bte:call-apis:query       'subject.umls:C0481980 AND predicate:TREATS AND object.semtypes:(acab OR anab OR cgab OR comd OR dsyn OR mobd OR neop)',
  bte:call-apis:query       'subject.umls:C0483573 AND predicate:TREATS AND object.semtypes:(acab OR anab OR cgab OR comd OR dsyn OR mobd OR neop)',
  bte:call-apis:query       'subject.umls:C0362338 AND predicate:TREATS AND object.semtypes:(acab OR anab OR cgab OR comd OR dsyn OR mobd OR neop)',
  bte:call-apis:query       'subject.umls:C0362339 AND predicate:TREATS AND object.semtypes:(acab OR anab OR cgab OR comd OR dsyn OR mobd OR neop)',
  bte:call-apis:query       'subject.umls:C0801917 AND predicate:TREATS AND object.semtypes:(acab OR anab OR cgab OR comd OR dsyn OR mobd OR neop)',
  bte:call-apis:query       'subject.umls:C0362340 AND predicate:TREATS AND object.semtypes:(acab OR anab OR cgab OR comd OR dsyn OR mobd OR neop)',
  bte:call-apis:query       'subject.umls:C0941434 AND predicate:TREATS AND object.semtypes:(acab OR anab OR cgab OR comd OR dsyn OR mobd OR neop)',
  bte:call-apis:query       'subject.umls:C0550679 AND predicate:TREATS AND object.semtypes:(acab OR anab OR cgab OR comd OR dsyn OR mobd OR neop)',
  bte:call-apis:query       'subject.umls:C0800007 AND predicate:TREATS AND object.semtypes:(acab OR anab OR cgab OR comd OR dsyn OR mobd OR neop)'
  bte:call-apis:query     ],
  bte:call-apis:query     scopes: []
  bte:call-apis:query   },
  bte:call-apis:query   method: 'post',
  bte:call-apis:query   timeout: 50000,
  bte:call-apis:query   headers: { 'User-Agent': 'BTE/dev Node/v18.16.1 darwin' }
  bte:call-apis:query } +1ms
  bte:call-apis:query query success, transforming hits->records... +373ms
  bte:api-response-transform:index api name BioThings SuppKG API +45s
  bte:api-response-transform:index api tags: drug,disease,association,chemical,metadata,query,translator,biothings +0ms
  bte:call-apis:query Successful POST https://biothings.ncats.io/suppkg (18 IDs): SmallMolecule > treats > Disease (obtained 0 records, took 372ms) +46ms
  bte:call-apis:query query completes. +0ms
  bte:call-apis:query Total number of records returned for this query is 0 +0ms
```

</p>
</details>

**Comment by rjawesome** [comment]
I'm working on fixing the Biothings extra functionality

**Comment by colleenXu** [comment]
@rjawesome @tokebe 

I think Jackson wants to get that working in JS rather than JQ. Check with them...

**Comment by tokebe** [comment]
@rjawesome @colleenXu Sorry, based on [this](https://github.com/biothings/biothings_explorer/issues/592#issuecomment-1489525088) it makes sense to me to have that fix made on the JQ side...I'm starting to think it actually makes sense to use JQ for Biothings instead of the JS transformer, as it adds another example case for people to work off of in the future (and having reviewed the JQ strings, it's actually quite easy to compare functionality to the JS implementation). 

As for performance (concern brought up by @newgene), it appears `node-jq` uses an actual binary of JQ rather than some form of JS implementation of JQ, so I'd be willing to bet performance would very much rival that of a current JS transformer.

@rjawesome please go ahead with the Biothings fix in JQ. Please note that I've pushed a commit to your branch which moves the JQ strings to their own files and reformats the code a little to make TypeScript happier, so you'll have to pull that to your local before making your changes.

**Comment by rjawesome** [comment]
Should be fixed. When edge.query_operation is serialized to JSON, it uses the actual variables rather than the getters from the class which have "_" (ie. _method instead of method) at the beginning, which was causing some issues.

**Comment by newgene** [comment]
@tokebe @rjawesome sounds good to me! JQ and native-JS are just two options we have, if performance is not an issue, I am fine to use either option, whichever is more convenient or efficient for us.

**Comment by newgene** [comment]
@tokebe @rjawesome @colleenXu We did not touch it at the meeting, but I want to bring your attention of our current x-bte transformation metadata syntax:

```
          transformer:
             wrap_jq: 
                "$ref": "#/components/x-bte-transformers/t1"
            pair_jq: 
                "$ref": "#/components/x-bte-transformers/t2"
```
as mentioned in https://github.com/biothings/api-respone-transform.js/pull/38#issuecomment-1329988691

If we want to change this syntax a bit, now it's the time.

One related note is I do think we should still add the `jmespath` support (later, not in this PR). With that in mind, I do think it's better to use `transformer.jq.{wrap|pair|}` syntax, so it's cleaner to add another transformer like `transformer.jmespath.*`. Your thoughts?

**Comment by tokebe** [comment]
Agreed, I think that makes sense.

**Comment by tokebe** [comment]
@rjawesome Looks like the bug with CTD is due to a case mismatch somewhere. 


<details>
<summary>This example drops all records</summary>

```
{
  "message": {
    "query_graph": {
      "nodes": {
        "n1": {
          "ids": [
            "KEGG.PATHWAY:hsa05323"
          ],
          "categories": [
            "biolink:Pathway"
          ]
        },
        "n2": {
          "categories": [
            "biolink:Gene"
          ]
        }
      },
      "edges": {
        "e1": {
          "subject": "n1",
          "object": "n2",
          "predicate": "biolink:has_participant"
        }
      }
    }
  }
}
```

</details>

<details>
<summary>While this one does not</summary>

```
{
  "message": {
    "query_graph": {
      "edges": {
        "e01": {
          "subject": "n0",
          "object": "n1",
          "predicates": [
            "biolink:related_to"
          ]
        }
      },
      "nodes": {
        "n0": {
          "ids": [
            "MESH:C006303"
          ],
          "categories": [
            "biolink:SmallMolecule"
          ]
        },
        "n1": {
          "categories": [
            "biolink:Gene"
          ]
        }
      }
    }
  }
}
```

</details>

Changing `ctd.jq` to use `ascii_downcase` instead of `ascii_upcase` flips it -- the first works, while the second does not. 

Update: It looks like the problem is somehow Record node curies becoming `undefined` as a result of the transformation, which downstream leads to the records not matching expectation, and thus being dropped.

**Comment by tokebe** [comment]
Update: Localized it. The Curie prefix is being `ascii_upcase`'d, but so is the ID -- which causes a mismatch. This should be changed to only apply to the prefix and not the ID. 

@rjawesome I'll handle this one.

Update: I understand the problem now -- CTD responds with basically all lowercase, regardless of whether it should be lowercase or not. So in example 1 from my above comment, the relevant ID should be `KEGG.PATHWAY:hsa05323`, but the current code changes it to `KEGG.PATHWAY:HSA05323`. Meanwhile, the second example should have `MESH:C006303` and CTD responds with `MESH:c006303`, but the current code then corrects it.

It appears we're going to have to either handle it on a prefix-by-prefix basis, or we'll have to get this fixed on CTD's end. @colleenXu I'd like your input on what we should do to move forward here.

**Comment by colleenXu** [comment]
@tokebe 

From what I can tell, [KEGG.PATHWAY IDs should have lower-case letters](https://bioregistry.io/registry/kegg.pathway) like `KEGG.PATHWAY:rsk00410` so the CTD's response is technically correct. 

I can see what you mean though. [MESH IDs should start with upper-case `C` or `D`](https://bioregistry.io/registry/mesh). But CTD's response has lower-case letters for those IDs...

---

I get the sense that this ID-formatting is not something that we should expect the CTD team to fix for their API. 

But I'm not sure whether BTE handling needs a small tweak (just for CTD-post-processing) vs something more involved (how BTE handles ID-letter-cases more generally).

**Comment by colleenXu** [comment]
@tokebe 

Quick note: both examples you give [here](https://github.com/biothings/biothings_explorer/issues/489#issuecomment-1728375525) seem to work on the main branch. I guess that code results in different behavior from the JQ code?

**Comment by colleenXu** [comment]
@tokebe

I'm overthinking about IDs. Recording my musings below:

<details><summary>(1) Are we on the same page?</summary>
<p>

I thought that within BTE, node/bioentity ID-values should be in their namespace's letter-case-format (which I guess is defined in [bioregistry](https://bioregistry.io/)). Does that sound right?
* however, I'm not sure if BTE handles these ID-value formats in a consistent way/stored in 1 place (maybe it's sprinkled into various api-transforms?)
* And I dunno how stable these formats are or how easy this would be to maintain. (reminds me of #591)
* I'm not clear on Translator's standards for ID-value format
    * while I get the sense that we're supposed to follow the [bioregistry](https://bioregistry.io/) for ID-value format...
    * it can be confusing to find the ID-value format by namespace-name/prefix because the biolink-model often uses namespace-name/prefixes that are different from bioregistry's
    * an example of this confusion is https://github.com/biolink/biolink-model/issues/1366 (do the ID-values start with `PMC` or not?)

    
I think it may be normal for BTE to adjust IDs for sub-queries (templating) and do post-processing on responses (jq, jmespath?) to handle the different ID-value formats that knowledgebase APIs may use...does that sound right? 
* When the knowledgebase is external to Translator and our lab (like CTD), I don't think we can expect their team to change their databases/APIs (but maybe it's okay to bring up to them). Especially when they're a popular resource with a lot of different users that'd be affected...

</p>
</details> 


<details><summary>(2) it seems uncommon for IDs to have lower-case letters. so maybe we could have a default all-caps...</summary>
<p>

While I didn't review all namespaces we use, I only had two come to mind immediately:
* `KEGG.PATHWAY:rsk00410`, as mentioned above: https://bioregistry.io/registry/kegg.pathway
* `DBSNP:rs121909098`: start with `rs`, https://bioregistry.io/registry/dbsnp

But I can think of a lot more examples of all-caps IDs:
* `MESH` IDs start with `C` or `D`: https://bioregistry.io/registry/mesh
* `KEGG.COMPOUND` IDs start with `C` (weird): https://bioregistry.io/registry/kegg.compound
* UMLS IDs start with `C`: https://bioregistry.io/registry/umls
* CHEMBL IDs start with `CHEMBL`: https://bioregistry.io/registry/chembl.compound
* UniProtKB IDs uses upper-case letters: https://bioregistry.io/registry/uniprot
* WIKIPATHWAY IDs start with `WP`: https://bioregistry.io/registry/wikipathways
* REACT (reactome) IDs use upper-case letters: https://bioregistry.io/registry/reactome

Maybe the "keep track of ID-value formats by namespace" would be easier if we designed BTE to keep a default ID-value format of all-caps, except in certain cases. 

We'd want to have super-good documentation in code + docs of this design decision though...

</p>
</details>


<details><summary>(3) Is something like this a concern with ID/namespace-prefixes too?</summary>
<p>

Maybe not, because we strip the prefix off of IDs during post-processing? 
* We do have a mix of namespace/ID-prefix formats, from all-caps to all-lower (ncats.bioplanet) to a mix of both (UniProtKB) 

But this does remind me of:
* #640, although that's a case of NodeNorm response (not knowledge API response)
* #591, although that's focused on the presence/absence of prefix on IDs...


</p>
</details>

**Comment by rjawesome** [comment]
> With that in mind, I do think it's better to use transformer.jq.{wrap|pair|} syntax, so it's cleaner to add another transformer like transformer.jmespath.*. Your thoughts?

Done.

**Comment by newgene** [comment]
@rjawesome @tokebe there is one more thing I want to bring up, and you can see if you like to bring it to bte (now or later):

In the jmespath implementation at biothings.api [here](https://github.com/biothings/biothings.api/blob/7b83617c40c2f12513a7db6842e5d14adf694039/biothings/web/query/formatter.py#L400) (you can just look at the doc string for an example), we introduced an optional `root_field`, so `jmespath` parameter can be passed as `<root_field>|<actual jmespath query>`. If `<root_field>` is empty or `.`, the transformation will still apply to the root. The benefit is when you only need to transform a nested field, this allows you to simplify the transformation query (don't have to figure out taking value out and put it back, which can be tricky to express in jq or jmespath, while it's relatively easy to handle in a generic js or python helper function. In biothings.api we have [a dictionary traversing helper function](https://github.com/biothings/biothings.api/blob/7b83617c40c2f12513a7db6842e5d14adf694039/biothings/utils/common.py#L896) to handle it).

Just want to put it here to hear your thoughts. It does not impact the current PR-merging, unless you think it's trivial to add in the current implementation.

**Comment by rjawesome** [comment]
JQ has the ``|=`` operator which should cover the case that you are saying. (it will pass the previous expression to the filter and put the following expression back in the field represented by the previous expression)

For example, if you wanted to add "a" at the end of some field
.field.this_is_an_array[].string_field |= . + "a"
(note that intermediate arrays have to be prefixed with "[]" to ensure that jq unpacks them)

<details><summary>Example input</summary>{

"field": {
"this_is_an_array": [

{"string_field":"b"},
{"string_field":"c"}
]

}


}</details>

<details><summary>Example output</summary>{

"field": {
"this_is_an_array": [

{"string_field":"ba"},
{"string_field":"ca"}
]

}


}</details>

**Comment by newgene** [comment]
@rjawesome that's nice. for a comparison, how you would do it in jq if you want to filter array nested field for only `string_field==b`?

**Comment by rjawesome** [comment]
Since the list filter has been created, the following should work (filtering for string_field==a)
.field.this_is_an_array |= list_filter_any(["string_field:a"]) -- if nested it could be like a.b.string_field:a and more conditions can be added to the array

For other besides ==, the list_filter function could be updated or the .[] operator could be used
.field.this_is_an_array |= [.[] | if .string_field == "a" then . else empty end] 
here the operator could be changed from == to something else

**Comment by colleenXu** [comment]
Follow-up: @tokebe @rjawesome 

I think the [BioThings APIs extra functionality](https://github.com/biothings/biothings_explorer/issues/592) is now working, yay!

I tested using the latest code with [Rohan's fixes](https://github.com/biothings/biothings_explorer/issues/489#issuecomment-1724667532) + [the testing situation I described in an earlier comment](https://github.com/biothings/biothings_explorer/issues/489#issuecomment-1720134312). 

I looked at the console logs to check that the functionality was working...

<details><summary>Example queries after setting up the testing situation</summary>
<p>

Checking that batch-querying IDs works alongside the "IDs don't need to be at the end of the query string of the BioThings response"


SmallMolecule Fluconazole, (r)-dithiolane-3-pentanoic acid -> Disease. Both should have edges to ulcerative colitis

```
{
    "message": {
        "query_graph": {
            "nodes": {
                "n0": {
                    "ids": ["UMLS:C0016277", "UMLS:DC0023791"],
                    "categories": ["biolink:SmallMolecule"]
                },
                "n1": {
                    "categories": ["biolink:Disease"]
                }
            },
            "edges": {
                "e01": {
                    "subject": "n0",
                    "object": "n1"
                }
            }
        }
    }
}
```

Disease Diverticulosis, muscular dystrophy -> SmallMolecule. 
```
{
    "message": {
        "query_graph": {
            "nodes": {
                "n0": {
                    "ids": ["UMLS:C1510475", "UMLS:C0026850"],
                    "categories": ["biolink:Disease"]
                },
                "n1": {
                    "categories": ["biolink:SmallMolecule"]
                }
            },
            "edges": {
                "e01": {
                    "subject": "n0",
                    "object": "n1"
                }
            }
        }
    }
}
```

</p>
</details>

**Comment by colleenXu** [comment]
Current situation, as discussed by Jackson @tokebe and I:

1. The CTD bug (convo starts [here](https://github.com/biothings/biothings_explorer/issues/489#issuecomment-1720118335)) is a regression in behavior, so we want to fix it. Jackson recently pushed commits that should address this so ➡️ next step is for me to test
    * I made a [vague post earlier](https://github.com/biothings/biothings_explorer/issues/489#issuecomment-1740082515) saying that both operations work w/ the current main branch code, but don't seem to work in the JQ branch ([as Jackson pointed out](https://github.com/biothings/biothings_explorer/issues/489#issuecomment-1728375525))
2. the [identifier-case / "lower case letters in IDs for some namespaces"](https://github.com/biothings/biothings_explorer/issues/489#issuecomment-1740106355) situation is separate and just being exposed by this api-response-transform module work. So I'll open a separate issue for this. 

<details><summary>some notes</summary>
<p>

* we're in agreement that this likely affects all transformers of the api-response-transform module
* we're in agreement that it'll be easier to have a default (all-caps) and then have a config of IDs that don't follow that. So...I'll need to curate a list of the ID namespaces we currently use that have lower-case letters in their IDs...
* Jackson pointed out that the rest of BTE seems to be handling identifier-case fine (and I couldn't find any previous issues on this topic). So perhaps this is mostly limited to api-response-transform functionality right now...

</p>
</details> 

3. Jackson says it's difficult to write and test JQ strings for BTE's implementation. I haven't tried yet, but I agree that it's a learning curve to read JQ strings...what can we do to make this easier for us and future contributors? EDIT: I'm going to make a separate issue on this, so convo can continue there...
    * Jackson says this likely isn't a blocker to getting this set of JQ PRs done and deployed, but it is an issue going forward
    * Jackson suggested a "testing tool" to help with writing and testing 

---

And next steps once the CTD fix is confirmed to work:
1. update branches in both PRs (https://github.com/biothings/api-respone-transform.js/pull/38, https://github.com/biothings/smartapi-kg.js/pull/61 to latest main. Check that everything still works?
2. for external apis, try moving the jq stuff to the smartapi yaml and test that all that works, [as discussed previously](https://github.com/biothings/biothings_explorer/issues/489#issuecomment-1718298042)
3. then discuss deployment. Particuarly, we'll want to figure out the following: 
    * should Biothings handling stay as “jq in bte” and not in the smartapi yamls? 
    * how do we deploy the JQ PRs, when we'd like to have "jq in smartapi" for external apis?
        * related: we'll want to update CTD x-bte annotation, to take advantage of batch-querying functionality in the JQ PRs...

<details><summary>My ideas on deployment</summary>
<p>

We've floated the idea of "smartapi overrides" before 

But I wonder, if an API has "JQ in smartapi" and "JQ in BTE", does the "JQ in smartapi" take priority ("JQ in BTE" not used)? That's my interpretation of the logic in the api-response-transform's `src/index.ts` file...And if so, maybe there could be another deployment method:  
* deploy with "jq in bte" to all instances 
* Once that's done, merge a registry repo PR with "jq in smartapi". This will then be used by the external apis instead of "jq in bte" 
* We'll have a PR to remove the "jq in bte" for these external apis. We can deploy it with other features anytime

</p>
</details>

**Comment by colleenXu** [comment]
Deployment discussion with Jackson @tokebe and I today:

I'll do many of the current status / next steps from the previous post

* BioThings handling: keep as "jq in bte". Having it in smartapi yamls would be annoying (changing in every yaml vs 1 spot) 
* multi-stage deployment:
    * deploy to all instances ("jq in bte" gets used)
        * we may be able to deploy "jq in smartapi" immediately, since it should just be ignored by instances that don't have the JQ PRs deployed yet. we should be able to test this before starting the deployment process
    * then...x-bte changes related to the new functionality can then be deployed
        * CTD batch-querying being one example (changing q strings and supportBatch:true)
        * BioThings extra functionality: errr....it doesn't seem urgent / necessary to change existing annotations to use this
    * then...we can delete stuff to avoid dev confusion (can deploy w/ other features anytime)
        * the "jq in bte" for external apis won't be needed if it's in the smartapi yaml
        * the old javascript transformers could be removed

**Comment by colleenXu** [comment]
@tokebe 

For CTD, it looks like the latest commits fix the KEGG.PATHWAY operation, yay!

But...it looks like a new bug's been introduced for [batch-querying post-processing](https://github.com/biothings/biothings_explorer/issues/584). Now when the test setup + query is run, every starting ID is linked to every output ID which is incorrect.

<details><summary>info</summary>
<p>

There should be 18 results, according to the desired response in the [linked issue](https://github.com/biothings/biothings_explorer/issues/584).

However, now I'm getting 34 results. And genes that should be only linked to 1 starting ID are instead showing up as linked to two IDs ([original API response](http://ctdbase.org/tools/batchQuery.go?inputType=chem&inputTerms=C006303%7CD015250&inputTermSearchType=directAssociations&report=genes_curated&format=json)):
* NCBIGene:142 (PARP1) should only be linked to D015250 (Aclarubicin; PUBCHEM.COMPOUND:451415)
* NCBIGene:1080 (CFTR) and NCBIGene:10800 (CYSLTR1) should only be linked to C006303 (acivicin; PUBCHEM.COMPOUND:294641)
 

</p>
</details>

**Comment by tokebe** [comment]
Ah, I think I see where this issue is occurring, let me put together a fix.

**Comment by tokebe** [comment]
Additionally, from discussion with @colleenXu regarding the `$edge` variable, here's an example:

<details>
<summary>$edge example</summary>

```javascript
{
  query_operation: {
    _params: {
      format: "json",
      inputTermSearchType: "directAssociations",
      inputTerms: "{{ queryInputs | replPrefix('KEGG') }}",
      inputType: "pathway",
      report: "genes_curated",
    },
    _requestBody: undefined,
    _requestBodyType: undefined,
    _supportBatch: false,
    _useTemplating: true,
    _inputSeparator: undefined,
    _templateInputs: undefined,
    _batchSize: undefined,
    _method: "get",
    _pathParams: [
    ],
    _server: "http://ctdbase.org",
    _path: "/tools/batchQuery.go",
    _tags: [
      "translator",
      "ctd",
    ],
    _transformer: {
      wrap_jq: undefined,
      pair_jq: undefined,
    },
  },
  association: {
    input_id: "KEGG.PATHWAY",
    input_type: "Pathway",
    output_id: "NCBIGene",
    output_type: "Gene",
    predicate: "has_participant",
    qualifiers: undefined,
    source: undefined,
    api_name: "CTD API",
    smartapi: {
      id: "0212611d1c670f9107baf00b77f0889a",
      meta: {
        date_created: "2023-03-14T06:06:50.582062+00:00",
        last_updated: "2023-05-22T07:04:00.328952+00:00",
        url: "https://raw.githubusercontent.com/NCATS-Tangerine/translator-api-registry/master/CTD/smartapi.yaml",
        username: "colleenXu",
      },
    },
    "x-translator": {
      component: "KP",
      team: [
        "Service Provider",
      ],
      infores: "infores:ctd",
    },
    apiIsPrimaryKnowledgeSource: true,
  },
  response_mapping: {
    has_participant: {
      NCBIGene: "data.GeneID",
      input_name: "data.PathwayName",
    },
  },
  tags: [
    "translator",
    "ctd",
  ],
  reasoner_edge: {
    id: "e1",
    predicate: undefined,
    subject: {
      id: "n1",
      category: [
        "biolink:Pathway",
      ],
      expandedCategories: [
        "Pathway",
      ],
      equivalentIDsUpdated: false,
      curie: [
        "KEGG.PATHWAY:hsa05323",
      ],
      is_set: undefined,
      expanded_curie: {
        "KEGG.PATHWAY:hsa05323": [
          "KEGG.PATHWAY:hsa05323",
        ],
      },
      entity_count: 1,
      held_curie: [
      ],
      held_expanded: {
      },
      constraints: undefined,
      connected_to: {
      },
      equivalentIDs: {
        "KEGG.PATHWAY:hsa05323": {
          primaryID: "KEGG.PATHWAY:hsa05323",
          equivalentIDs: [
            "KEGG.PATHWAY:hsa05323",
          ],
          label: "KEGG.PATHWAY:hsa05323",
          labelAliases: [
            "KEGG.PATHWAY:hsa05323",
          ],
          primaryTypes: [
            "Pathway",
          ],
          semanticTypes: [
            "Pathway",
          ],
          attributes: {
          },
        },
      },
    },
    object: {
      id: "n2",
      category: [
        "biolink:Gene",
      ],
      expandedCategories: [
        "Gene",
      ],
      equivalentIDsUpdated: false,
      curie: undefined,
      is_set: undefined,
      expanded_curie: {
      },
      entity_count: 0,
      held_curie: [
      ],
      held_expanded: {
      },
      constraints: undefined,
      connected_to: {
      },
    },
    expanded_predicates: undefined,
    qualifier_constraints: [
    ],
    reverse: false,
    executed: false,
    logs: [
    ],
    records: [
    ],
  },
  input: {
    queryInputs: "hsa05323",
  },
  input_resolved_identifiers: {
    "KEGG.PATHWAY:hsa05323": {
      primaryID: "KEGG.PATHWAY:hsa05323",
      equivalentIDs: [
        "KEGG.PATHWAY:hsa05323",
      ],
      label: "KEGG.PATHWAY:hsa05323",
      labelAliases: [
        "KEGG.PATHWAY:hsa05323",
      ],
      primaryTypes: [
        "Pathway",
      ],
      semanticTypes: [
        "Pathway",
      ],
      attributes: {
      },
    },
  },
  original_input: {
    "KEGG.PATHWAY:hsa05323": "KEGG.PATHWAY:hsa05323",
  },
  filter: undefined,
}
```

</details>

**Comment by tokebe** [comment]
Just to follow-up: I think this should _probably_ be abstracted into a simpler interface for anybody writing a JQ string in their annotation. If so, that would be a blocking requirement prior to deployment. Any work on such an abstracted interface is pending further discussion between @colleenXu and myself, so I'll handle implementation if we move in that direction.

**Comment by tokebe** [comment]
@colleenXu CTD batch processing should now be fixed, please re-test and let me know if there are any other misbehaviors.

**Comment by colleenXu** [comment]
@tokebe @andrewsu @newgene 

<details><summary>Hmmm...I'd like to add some details from the discussion Jackson and I had yesterday (related to the 3 previous posts by Jackson). </summary>
<p>

I have a collection of saved posts from this issue and PRs, on how to actually write JQ stuff for BTE.

One of these posts says that the [JQ strings can use variables from BTE's internal representation of stuff ($edge)](https://github.com/biothings/api-respone-transform.js/pull/38#issuecomment-1314651390). 

Jackson and I discussed this:
* it seems like "pair" JQ-strings use info from $edge, and "wrap" doesn't
    * "pair" is defined in this [post](https://github.com/biothings/api-respone-transform.js/pull/38#issuecomment-1482128403) as "match each entry returned by API to it's curie in an object"
    * VS "wrap" which is "process the raw data from the API before the response mapping is applied"
* $edge is a pretty complicated object, and it's not always clear at a glance what its various fields store (unless a fully-fleshed out instance is pulled out during query-execution using breakpoints) 
* $edge is related to `BTEKGOperationObject` (I'm not clear on what this is, relative to kinds of Edges summarized [here](https://github.com/biothings/biothings_explorer/issues/379#issuecomment-1068057622) during the vocab refactor)

</p>
</details> 

But...I'm still unsure on whether this should be a blocker to deployment. Could it be saved for a next step/phase of BTE JQ work?

* how often is $edge info needed?
* how urgent / necessary is it to have "abstraction" during this phase of JQ work? Maybe it's okay to implement and use this work for a while first, to gather the info needed to build the "abstraction":
    * right now, it's unclear to me what info from $edge is needed for JQ-string-writing
    * there's also what format the $edge info will be in + what format it needs to stay in (in case JQ-work is actually changing the variables in $edge, but I dunno if any of our current code-strings do this?). 
* In the future, do we expect outside devs to write JQ-strings that involve $edge info (VS BTE devs always taking the lead on this)?

**Comment by tokebe** [comment]
- If we deploy, and then later restructure `$edge`, you have to go back and fix the jq strings on any annotation that uses it. Personally, I'm against that, but if getting JQ out the door is a higher priority than getting it out the door in a polished state, then I defer to @colleenXu @andrewsu @newgene.
- I expect `$edge` to be used in almost any annotation that uses the pair function, because it contains information that is often useful for the logic, as well as giving the edge inputs where an API might not be returning them, or returning them malformed. 
- I'd expect that the intent is for outside devs to occasionally use this, as it's a part of the annotation, which we already expect outside devs to be able to work on. I don't see how a non-abstracted, very dense edge object is desirable, even if BTE-only devs are the ones using it.

**Comment by tokebe** [comment]
From discussion with @andrewsu:

For now, table jq in smartapi -- don't move any jq strings to smartapi and don't advertise it as an option. Move forward with deploying code for jq in BTE (it's fine to deploy the jq in smartapi code, just don't start moving jq to smartapi). We can then come back later and revisit this for usability improvements and move forward with jq in smartapi.

**Comment by newgene** [comment]
I also agree with that plan. Starting with our own use first before we advertise it. The original motivation of this issue is to reduce the needs of bte's own code changes, mostly for us.

**Comment by colleenXu** [comment]
@tokebe 

I retested CTD and all expected behaviors are working, yay! 
* basic batch-querying behavior (follow-up to https://github.com/biothings/biothings_explorer/issues/489#issuecomment-1760024369)
* all operations (including [KEGG.PATHWAY](https://github.com/biothings/biothings_explorer/issues/489#issuecomment-1720118335))
* [pipe-delimited values, noted here](https://github.com/biothings/biothings_explorer/issues/489#issuecomment-1718298042) as part of this feature)

I made a post [here](https://github.com/biothings/biothings_explorer/issues/584#issuecomment-1765469320) on whether BTE will need more adjustments for CTD batch-querying support. However, I don't think this is necessarily a barrier to deploying "JQ in BTE". 

EDIT: I've double-checked [all APIs with "JQ in BTE"](https://github.com/biothings/biothings_explorer/issues/489#issuecomment-1718298042), and no regression in behavior....all are working as-expected. 

I also tested w/ and w/o the smartapi-kg PR, and the basic "JQ in BTE" behavior appeared identical...

**Comment by colleenXu** [comment]
Jackson @tokebe and I decided to deploy only https://github.com/biothings/api-respone-transform.js/pull/38 to address this issue

Details from our Slack DM convo:
* Jackson: "[smartapi-kg PR](https://github.com/biothings/smartapi-kg.js/pull/61) likely won't need significant changes in the future (everything needing changes that I'm aware of for future jq-in-smartapi stuff is in the api-response-transform)"
* My reasons to put off deploying the smartapi-kg PR right now:
    * If we deploy it now, then "minimal JQ in SmartAPI" is basically silently live. So if we want to deploy "better JQ in SmartAPI" in the future w/ an api-response-transform PR, we have to wait for that future PR to deploy to all instances before we can safely update the yamls (x-bte w/ JQ)
    * VS if we waited to deploy the smartapi-kg PR with the future api-response-transform PR, we can also deploy the updated yamls right away - instances that don't have the PRs yet won't be able to ingest the new fields and will just ignore it
* Jackson's reply: "You're right. I overlooked that, we should leave out smartapi-kg for now"

**Comment by tokebe** [comment]
Relevant changes deployed to Prod.
